### PR TITLE
[BEAM-1629] Init metrics/aggregators accumulators before traversing pipeline

### PR DESCRIPTION
**Pull request by aviemzur**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by aviemzur** [comment]
R: @staslev 
CC: @amitsela

**Comment by aviemzur** [comment]
Run Spark RunnableOnService

**Comment by aviemzur** [comment]
Retest this please.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/metrics/MetricsAccumulator.java [hunk 1]
```diff
@@ -46,21 +49,38 @@
...
-          SparkMetricsContainer initialValue = new SparkMetricsContainer();
-          instance = jsc.sc().accumulator(initialValue, "Beam.Metrics",
-              new MetricsAccumulatorParam());
-          if (checkpointDir.isPresent()) {
-            recoverValueFromCheckpoint(jsc, checkpointDir.get());
+          Optional<CheckpointDir> maybeCheckpointDir =
+              opts.isStreaming() ? Optional.of(new CheckpointDir(opts.getCheckpointDir()))
+                  : Optional.<CheckpointDir>absent();
+          Accumulator<SparkMetricsContainer> accumulator =
+              jsc.sc().accumulator(new SparkMetricsContainer(), "Beam.Metrics",
```
Comment (staslev) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/metrics/MetricsAccumulator.java]: Extract constant - `Beam.Metrics`.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/SparkRunner.java [hunk 2]
```diff
@@ -218,11 +185,12 @@ public void run() {
       final JavaSparkContext jsc = SparkContextFactory.getSparkContext(mOptions);
       final EvaluationContext evaluationContext = new EvaluationContext(jsc, pipeline);
 
+      initAccumulators(mOptions, jsc);
```
Comment (staslev) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/SparkRunner.java]: In streaming mode, seems like we call `initAccumulators(...)` both here and [here](https://github.com/apache/beam/pull/2171/files#diff-b68cfd0ea98c566fbe6024c78e89342cR83). 
While multiple calls cause no harm in this case, do we really need both?

I assume the latter is required to satisfy [this](https://github.com/apache/beam/pull/2171/files#diff-08c9c68e4e7d25b7acf16d32fc66d601R136), i.e., to make sure things have been initialised by the time the pipeline is translated and needs an `AggregatorsAccumulator` instance.

I think we can make `SparkRunner#run()`'s interaction with `SparkRunnerStreamingContextFactory` and the pipeline translation in particular, clearer (pipeline translation is hidden inside `SparkRunnerStreamingContextFactory#create()` which IMHO is not ideal).

Anyhow, this is out of the scope of this PR, but could be nice to look into sometime.

Comment (aviemzur) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/SparkRunner.java]: I'll add comments to explain why this is done.

**Comment by aviemzur** [comment]
Run Spark RunnableOnService

**Comment by aviemzur** [comment]
@staslev addressed your comments in my latest commits.

**Comment by aviemzur** [comment]
@staslev brought back `CompositeSource` as we discussed.
PTAL

**Comment by aviemzur** [comment]
Retest this please.

**Comment by aviemzur** [comment]
Retest this please.

**Comment by aviemzur** [comment]
Rebased on top of master.

**Comment by staslev** [comment]
Run Spark RunnableOnService

**Comment by aviemzur** [comment]
Rebased on top of master.

**Comment by aviemzur** [comment]
Run Spark RunnableOnService

**Comment by aviemzur** [comment]
Run Spark RunnableOnService

**Pull request closed without merging**